LCLSGeneral/LCLSGeneral/POUs/Diagnostics/FB_EtherCATDiag.TcPOU

```diff
@@ -0,0 +1,575 @@
+
+bSlaveCountError := (nSlaveCount <> nSlaveCountCfg) OR (nSlaveCount = 0);
+IF (bSlaveCountError AND NOT bSlaveCountErrorPrev) OR (NOT bSlaveCountError AND bSlaveCountErrorPrev) THEN
+	bSlaveCountErrorPrev := bSlaveCountError;
+	bDiagReq := TRUE; // slave count error change detected --> diag required
+END_IF
+
+IF (bAllFrameWcStatesOK AND NOT bAllFrameWcStatesOKPrev) OR (NOT bAllFrameWcStatesOK AND bAllFrameWcStatesOKPrev) THEN
+	bAllFrameWcStatesOKPrev := bAllFrameWcStatesOK;
+	bDiagReq := TRUE; // frame error change detected --> diag required
+END_IF
+
+IF (nMasterDevState <> nMasterDevStatePrev) THEN
+	M_CheckMasterDevState();
+	bDiagReq := TRUE; // devstate change detected --> diag required
+END_IF
+
+(* acyclic diag *)
+CASE iState OF
+0: (* IDLE *)
+	IF bDiagReq THEN				// diag requested
+		bDiagReq := FALSE;
+
+		IF sMasterNetID <> '' AND sMasterNetID <> '0.0.0.0.0.0' THEN
+			iState := iState + 1;	// execute diag
+			bBusy := TRUE;
+		ELSE
+			bError := TRUE;
+			iErrorID := 7;
+		END_IF
+
+		bEtherCATOK := FALSE;
+	ELSE
+		// check for changes in idle
+		IF (bSlaveCountError OR bMasterDevStateError OR NOT bAllFrameWcStatesOK) AND NOT (arrSlaveInfo[iDiagIndex[0]].bDiagData) THEN
+			bEtherCATOK := FALSE;
+			bDiagReq := TRUE;//new error --> diag requested    
+		ELSIF (bSlaveCountError AND NOT bSlaveCountErrorPrev) OR (NOT bSlaveCountError AND bSlaveCountErrorPrev) THEN
+			bSlaveCountErrorPrev := bSlaveCountError;
+			bEtherCATOK := FALSE;
+			bDiagReq := TRUE;// slave count error change detected --> diag required
+		ELSIF (nMasterDevState <> nMasterDevStatePrev) THEN
+			bEtherCATOK := FALSE;
+			bDiagReq := TRUE;// devstate change detected --> diag required
+		ELSIF (bAllFrameWcStatesOK AND NOT bAllFrameWcStatesOKPrev) OR (NOT bAllFrameWcStatesOK AND bAllFrameWcStatesOKPrev) THEN
+			bAllFrameWcStatesOKPrev := bAllFrameWcStatesOK;
+			bEtherCATOK := FALSE;
+			bDiagReq := TRUE;// frame error change detected --> diag required
+		ELSIF (bSlaveCountError OR bMasterDevStateError OR NOT bAllFrameWcStatesOK OR arrSlaveInfo[iDiagIndex[0]].bDiagData) THEN
+			bEtherCATOK := FALSE;
+		ELSE
+			bEtherCATOK := TRUE;
+		END_IF
+	END_IF
+
+1: (* get adresses *)
+	M_GetSlaveAdresses();
+
+2: (* get states *)
+	M_GetSlaveStates();
+
+3:	(* get topology data length *)
+	M_GetTopoDataLen();
+
+4:	(* get topology data *)
+	M_GetTopoData();
+
+5:	(* scan slaves *)
+	M_ScanSlaves();
+
+6:	(* get identity *)
+	M_GetSlaveIdentity();
+
+7:	(* get name *)
+	M_GetSlaveName();
+
+8:	(* get scanned name *)
+	M_GetScannedSlaveName();
+
+10: (* Log diagnostics *)
+    (* I can't get the fbJsonDataType to actually convert the slave info 
+    structures. I just get nulls. Either I am doing this wrong, or when 
+    the symbol parser encounters datatypes it can't deal with it nulls
+    the whole thing. I'll keep this code commented out until someone figures
+    out how to deal with parsing the slave structs into the json payload *)
```

Comment:
Interesting... what do you mean global?

Reply:
Yeah, a global type in the  **_type system_**

Reply:
The other thing to consider is a pytmc of the array... it might be nice to make stats on diagnostic message payloads, but with EPICS+Archiver+Grafana we'll have that anyways.

Reply:
Ugh, you said it...

Reply:
I guess it wasn't too painful to add stuff the last time so maybe this won't be so bad. I'll give it a try.

Reply:
Last time being when you first added logging to general.

Reply:
How can you tell?

Reply:
It says so. Something about a structure in a Beckhoff library that can't be made into a global type because it's closed source.